### ISSUE TITLE
Extra tasks in freeze cgroup result in bad parasite management

### ISSUE THREAD

**Issue by jcarnus** [xemul/criu#79]
Hi
I tried to make a  "lx-checkpoint -n ...." or the "criu dump -t 8068" on a frozen container and the command rever return.

Here the logs

(00.000048) Probing sock diag modules
(00.000116) Done probing
(00.000123) ========================================
(00.000130) Dumping processes (pid: 8068)
(00.000132) ========================================
(00.000134) Running pre-dump scripts
(00.000187) Found anon-shmem device at 5
(00.000200) Reset 10168's dirty tracking
(00.000242)  ... done
(00.000275) Dirty track supported on kernel
(00.000337) Found task size of 7ffffffff000
(00.000352) irmap: Searching irmap cache in work dir
(00.000366) No irmap-cache image
(00.000370) irmap: Searching irmap cache in parent
(00.000412) irmap: No irmap cache
(00.000428) cpu: fpu:1 fxsr:1 xsave:1
(00.000506) vdso: Parsing at 7ffdaf9ec000 7ffdaf9ee000
(00.000512) vdso: PT_LOAD p_vaddr: 0
(00.000515) vdso: DT_HASH: 120
(00.000517) vdso: DT_STRTAB: 268
(00.000519) vdso: DT_SYMTAB: 160
(00.000521) vdso: DT_STRSZ: 5e
(00.000522) vdso: DT_SYMENT: 18
(00.000524) vdso: nbucket 3 nchain b bucket 7ffdaf9ec128 chain 7ffdaf9ec134
(00.000532) vdso: rt [vdso] 7ffdaf9ec000-7ffdaf9ee000 [vvar] 7ffdaf9ea000-7ffdaf9ec000
(00.000592) Writing image inventory (version 1)
(00.000624) Add pid ns 1 pid 10168
(00.000634) Add net ns 2 pid 10168
(00.000641) Add ipc ns 3 pid 10168
(00.000647) Add uts ns 4 pid 10168
(00.000653) Add mnt ns 5 pid 10168
(00.000660) Add user ns 6 pid 10168
(00.000663) cg: Dumping cgroups for 10168
(00.000691) cg:  `- New css ID 1
(00.000696) cg:     `- [blkio] -> [/]
(00.000698) cg:     `- [cpu,cpuacct] -> [/]
(00.000700) cg:     `- [cpuset] -> [/]
(00.000702) cg:     `- [devices] -> [/]
(00.000704) cg:     `- [freezer] -> [/]
(00.000706) cg:     `- [hugetlb] -> [/]
(00.000708) cg:     `- [memory] -> [/]
(00.000710) cg:     `- [name=systemd] -> [/]
(00.000712) cg:     `- [net_cls,net_prio] -> [/]
(00.000713) cg:     `- [perf_event] -> [/]
(00.000715) cg: Set 1 is criu one

**Comment by xemul** [comment]
Try using the --freeze-cgroup option. Does it help?

**Comment by jcarnus** [comment]
Sory, using criu v1.7-177, there is no --freeze-cgroup option no ?

**Comment by xemul** [comment]
What does -777 mean? :) The --freeze-cgroup appeared in 1.7

**Comment by jcarnus** [comment]
root@debian:~# criu --version
Version: 1.7
GitID: v1.7-177-g628a488


And 
/root/criu/criu dump -t 1259 --freeze-cgroups /sys/fs/cgroup/freezer/test
or
/root/criu/criu dump -t 1259 --freeze-cgroups
shows me the usage :)

**Comment by xemul** [comment]
I didn't tell the option name was "--freeze-cgroupS", I told it was "--freeze-cgroup".

**Comment by jcarnus** [comment]
Yes sorry fot that
Now
root@debian:~# /root/criu/criu dump -t 6837 --freeze-cgroup /sys/fs/cgroup/freezer/lxc/101 --enable-external-masters --ext-mount-map auto
Error (parasite-syscall.c:922): Task is in unexpected state: d7f
Error (cr-dump.c:1304): Can't cure (pid: 6837) from parasite
Error (cr-dump.c:1631): Dumping FAILED.

**Comment by xemul** [comment]
Show full log please.

**Comment by jcarnus** [comment]
Here we go : http://pastebin.com/C1JA13y2

**Comment by xemul** [comment]
OK, so we're detaching from parasites and the root task gets:

   (00.098310) Error (parasite-syscall.c:922): Task is in unexpected state: d7f
   (00.098316) Error (cr-dump.c:1304): Can't cure (pid: 6837) from parasite

The state is stopped but stopsignal is SIGPIPE %)
Does it happen in 100% cases when you try to dump the container?

**Comment by jcarnus** [comment]
I just tried 6 more time
So we are on 8/8 failed
Do you need more info to solve this ?

**Comment by xemul** [comment]
diff --git a/arch/x86/crtools.c b/arch/x86/crtools.c
    index 1d50c07..0960f3c 100644
    --- a/arch/x86/crtools.c
    +++ b/arch/x86/crtools.c
    @@ -531,6 +531,7 @@ int sigreturn_prep_fpu_frame(struct rt_sigframe *sigframe, fpu_state_t *fpu_stat
     
     int ptrace_set_breakpoint(pid_t pid, void *addr)
     {
    +#if 0
     	int ret;
     
     	/* Set a breakpoint */
    @@ -556,10 +557,13 @@ int ptrace_set_breakpoint(pid_t pid, void *addr)
     	}
     
     	return 1;
    +#endif
    +	return 0;
     }
     
     int ptrace_flush_breakpoints(pid_t pid)
     {
    +#if 0
     	/* Disable the breakpoint */
     	if (ptrace(PTRACE_POKEUSER, pid,
     			offsetof(struct user, u_debugreg[DR_CONTROL]),
    @@ -567,6 +571,7 @@ int ptrace_flush_breakpoints(pid_t pid)
     		pr_perror("Unable to disable the breakpoint for %d", pid);
     		return -1;
     	}
    +#endif
     
     	return 0;
     }


Would you try with this patch, please? (Tabs probably got corrupted with spaces, use -l option when applying).

**Comment by jcarnus** [comment]
After path,
function is
int ptrace_set_breakpoint(pid_t pid, void *addr)
{
#if 0
        int ret;

        /* Set a breakpoint */
        if (ptrace(PTRACE_POKEUSER, pid,
                        offsetof(struct user, u_debugreg[DR_FIRSTADDR]),
                        addr)) {
                pr_perror("Unable to setup a breakpoint into %d", pid);
                return -1;
        }

        /* Enable the breakpoint */
        if (ptrace(PTRACE_POKEUSER, pid,
                        offsetof(struct user, u_debugreg[DR_CONTROL]),
                        X86_DR_LOCAL_ENABLE(DR_FIRSTADDR))) {
                pr_perror("Unable to enable the breakpoint for %d", pid);
                return -1;
        }

        ret = ptrace(PTRACE_CONT, pid, NULL, NULL);
        if (ret) {
                pr_perror("Unable to restart the  stopped tracee process %d", pid);
                return -1;
        }

        return 1;
#endif
   return 0;
}


make clean && make
After failed
Error log:http://pastebin.com/45v4grWg

**Comment by avagin** [comment]
```
(00.008404) Waiting for 5397 to trap
(00.008425) Daemon 5397 exited trapping
(00.008439) Sent msg to daemon 6 0 0
(00.008467) 6769 was trapped
```

A wrong process was trapped (6769 instead of 5397). This process lives in a freezer cgroup, but it isn't in a container process tree.

**Comment by avagin** [comment]
@jcarnus Could you try out this patch: https://github.com/avagin/criu/commit/585e7ba6d4bd62371a9791b5fe4d222359f5a60e

**Comment by jcarnus** [comment]
Sorin,
Still failed
Log here : http://pastebin.com/mZ8GbYh8

**Comment by jcarnus** [comment]
I cannot compile : 
  PBCC     protobuf/google/protobuf/descriptor.pb-c.c
google/protobuf/descriptor.proto:1:1: Expected top-level statement (e.g. "message").
make[1]: *** No rule to make target 'protobuf/google/protobuf/descriptor.pb-c.c', needed by 'protobuf/google/protobuf/descriptor.pb-c.d'.  Stop.
make[1]: Leaving directory '/root/avh/criu2'
Makefile:210: recipe for target 'protobuf' failed
make: *** [protobuf] Error 2

**Comment by jcarnus** [comment]
I managed to compile.
Failed but due to another error
Warn  (arch/x86/crtools.c:133): Will restore 5742 with interrupted system call
Warn  (arch/x86/crtools.c:133): Will restore 5816 with interrupted system call
Warn  (arch/x86/crtools.c:133): Will restore 5837 with interrupted system call
Warn  (fsnotify.c:306): fsnotify:               Detected FS_EVENT_ON_CHILD bit in mask (will be ignored on restore)
tar: ./dbus/system_bus_socket: socket ignored
tar: ./rpcbind.sock: socket ignored
tar: ./log: socket ignored
Error (namespaces.c:815): Namespaces dumping finished with error 8390015
Error (cr-dump.c:1631): Dumping FAILED.


Full log
http://demo.ovh.eu/fr/9eedc65666b3849f338987173bc67736/

**Comment by avagin** [comment]
It's the same problem in another place. I'm going to fix this issue on Monday. Thank you for the help.

**Comment by avagin** [comment]
@jcarnus Could you try out this patch https://lists.openvz.org/pipermail/criu/2015-October/023108.html?

**Comment by avagin** [comment]
Here is the same patch https://github.com/avagin/criu/commit/5f34093e2fc151857b7f622b9843a3cef4d0072d

**Comment by jcarnus** [comment]
Nop
The log http://pastebin.com/NxDbiQtg

Do i need to cumulate patch ?

**Comment by avagin** [comment]
@jcarnus looks like you try to dump processes which are not in the specifies freezer cgroup.

waitpid returned -1 (errno = 10 (ECHLD))

Could you show output for "ps axf" "cat /sys/fs/cgroup/freezer/PATH_TO_CT_GROUP"?

**Comment by jcarnus** [comment]
You have everythin here http://demo.ovh.eu/fr/e7ad9837192683ceeb3b4e05bf2f660b/

Command :
root@debian:~# lxc-start -n 101
root@debian:~/criu# lxc-freeze -n 101
root@debian:~/criu# lxc-info -n 101
Name:           101
State:          FROZEN
PID:            1755
IP:             192.168.0.120
CPU use:        1.17 seconds
BlkIO use:      20.57 MiB
Memory use:     13.96 MiB
KMem use:       0 bytes
Link:           veth101i0
 TX bytes:      798 bytes
 RX bytes:      5.73 KiB
 Total bytes:   6.50 KiB
criu dump -t 1755 --freeze-cgroup /sys/fs/cgroup/freezer/lxc/101 --enable-external-masters --ext-mount-map auto -v4 -o criu.log

**Comment by avagin** [comment]
I need a content of the "tasks" file for the freeze cgroup.

**Comment by avagin** [comment]
Why do you have so many lxc-attach processes?

They live in the container's group.
```(00.014586) Daemon 1755 exited trapping
(00.014598) Sent msg to daemon 6 0 0
pie: __fetched msg: 6 0 0
pie: 1: new_sp=0x7f0cfc59c008 ip 0x7f0cfbc73291
(00.015431) 2948 was trapped
```

```
2948 ?        D      0:00 lxc-attach -n 101 -- df -P -B 1 /
```

**Comment by jcarnus** [comment]
I rebooted the VM (Proxmox 4)
Started CT 101; froze it and info it to get the piD
Here the log
http://demo.ovh.eu/fr/214720d65b9c8973ff66313ad5d5d8f9/

Task file coming from /sys/fs/cgroup/freezer/lxc/101/tasks seems always empty
You have ps log before and after criu dump

**Comment by avagin** [comment]
> Task file coming from /sys/fs/cgroup/freezer/lxc/101/tasks seems always empty

All container's processes should live in this group.

@tych0 Could you help us?

**Comment by tych0** [comment]
[Sorry for the delay, I wasn't subscribed to criu's github notices.
I think I've just subscribed, so hopefully I can chime in in the
future.]

On Tue, Oct 27, 2015 at 10:50:54PM -0700, Andrew Vagin wrote:
> > Task file coming from /sys/fs/cgroup/freezer/lxc/101/tasks seems always empty
> 
> All container's processes should live in this group.

Hmm. I don't see it is empty, but I do see that only the container's
init task lives in the freezer cgroup which also seems incorrect.

What version of liblxc1 and lxc are you using? What is your host
distro? If not ubuntu, are you using lxcfs?

**Comment by tych0** [comment]
I am traveling right now so can't debug in detail, but hopefully there will be some resolution here:
https://lists.linuxcontainers.org/pipermail/lxc-devel/2015-October/012635.html

Otherwise I'll try and look further when I get home Friday.

**Comment by jcarnus** [comment]
Ok so
Before starting the CT 101, I have no /sys/fs/cgroup/freezer/lxc/101 folder
But a /sys/fs/cgroup/freezer/taks contains a lofof iipds
Starting CT101 make /sys/fs/cgroup/freezer/lxc/101 appears
Inside tasks file contains like 20-25 pid (/sys/fs/cgroup/freezer/lxc/101/tasks)

With that, a criu dump never return 
Here log for tasks file and criu.log
http://demo.ovh.eu/fr/26ecbb74bba8e42bb0c1b0320d7959ff/

**Comment by tych0** [comment]
On Wed, Oct 28, 2015 at 06:35:58AM -0700, jcarnus wrote:
> Ok so
> Before starting the CT 101, I have no /sys/fs/cgroup/freezer/lxc/101 folder
> But a /sys/fs/cgroup/freezer/taks contains a lofof iipds
> Starting CT101 make /sys/fs/cgroup/freezer/lxc/101 appears
> Inside tasks file contains like 20-25 pid (/sys/fs/cgroup/freezer/lxc/101/tasks)
> 
> With that, a criu dump never return 
> Here log for tasks file and criu.log
> http://demo.ovh.eu/fr/26ecbb74bba8e42bb0c1b0320d7959ff/

Are you using the -s option to `lxc-checkpoint`? Looks like you're
not, which is probably not what you want.

**Comment by jcarnus** [comment]
I use a criu compiled with patch. I am not using lxc-checkpoint. M'y goal is to chexkpoint à frozen lxc ct

**Comment by avagin** [comment]
I've fixed a few bugs in my patch. There is a new version: https://github.com/avagin/criu/commit/4999c10e139147bae74371ada9229e818a173b95

**Comment by jcarnus** [comment]
Ok so now I am able to make a dump with success. I got CT stopped and a lot of *.img file.
If i tried a restore, I have the following error : 
(00.006633) Error (mount.c:2393): mnt: No mapping for /sys/fs/cgroup/perf_event mountpoint

What does it mean ? Why I am able to dump if restore is not possible ?

**Comment by avagin** [comment]
Which commands do you use to dump and restore a container?
You can try to add --ext-mount-map auto to both commands.

**Comment by jcarnus** [comment]
I use the lxc-checkpoint to dump and to restore
The code here https://github.com/lxc/lxc/blob/master/src/lxc/criu.c shows that flags --ext-mount-map its always added

**Comment by avagin** [comment]
@tych0 we need your help again.

**Comment by tych0** [comment]
On Fri, Oct 30, 2015 at 05:47:38AM -0700, jcarnus wrote:
> I use the lxc-checkpoint to dump and to restore

> I use a criu compiled with patch. I am not using lxc-checkpoint. M'y
> goal is to chexkpoint à frozen lxc ct

Which is it?

If you're using lxc-checkpoint, which version of lxc is it? I think
you'll want 1.1.4 in order to have lxc-checkpoint pass --ext-mount-map
auto.

**Comment by jcarnus** [comment]
Sorry for the delay
I now have a lxc 1.1.4.
But I wanted to make a criu dump command and I am facing the same king of issue to dump
Here the log file 
http://pastebin.com/5LPysM8v

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 05:56:57AM -0800, jcarnus wrote:
> Sorry for the delay
> I now have a lxc 1.1.4.
> But I wanted to make a criu dump command and I am facing the same king of issue to dump
> Here the log file 
> http://pastebin.com/5LPysM8v

This is a different issue, but right now you need to disable seccomp
in order to c/r successfully (although hopefully you won't by 1.8).
Put:

lxc.seccomp =

in your container's config file somewhere.

**Comment by jcarnus** [comment]
Yes sorry, I forgot.
So able to checkpoint now
So  criu restore --ext-mount-map auto -o out1.log -vvvv failed with the following log
http://pastebin.com/N6aydhbJ

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 10:10:57AM -0800, jcarnus wrote:
> Yes sorry, I forgot.
> So able to checkpoint now
> So  criu restore --ext-mount-map auto -o out1.log -vvvv failed with the following log

I highly recommend you just use lxc-checkpoint to restore as well.

> http://pastebin.com/N6aydhbJ

Looks like your target host didn't have lxcfs?

**Comment by jcarnus** [comment]
Yes ok. Here the restore.log coming from lxc-checkpoint -r

http://demo.ovh.eu/fr/ee10394ce573db1cd0205baa1ad7515f/

What do you mean by lxcfs ? LXC Container is stored on a ZFS volume

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 10:21:59AM -0800, jcarnus wrote:
> Yes ok. Here the restore.log coming from lxc-checkpoint -r
> 
> http://demo.ovh.eu/fr/ee10394ce573db1cd0205baa1ad7515f/
> 
> What do you mean by lxcfs ? LXC Container is stored on a ZFS volume

Ok, now you've got a different error:

(00.107946) Error (net.c:983): Can't add interface veth100i0 to bridge (null): No such device

Did you copy the container's config file over as well before restoring
it?

**Comment by jcarnus** [comment]
I tryed on the same machine.
lxc-checkpount dump
lxc-stop
lxc-checkpoint restaure

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 10:42:15AM -0800, jcarnus wrote:
> I tryed on the same machine.
> lxc-checkpount dump
> lxc-stop
> lxc-checkpoint restaure

Can you paste the container's config file?

**Comment by jcarnus** [comment]
root@debian:~# cat /var/lib/lxc/100/config
lxc.arch = amd64
lxc.include = /usr/share/lxc/config/debian.common.conf
lxc.console = none
lxc.cgroup.devices.deny = c 5:1 rwm
lxc.tty = 0
lxc.environment = TERM=linux
lxc.utsname = pbx
lxc.cgroup.memory.limit_in_bytes = 536870912
lxc.cgroup.memory.memsw.limit_in_bytes = 1073741824
lxc.cgroup.cpu.cfs_period_us = 100000
lxc.cgroup.cpu.cfs_quota_us = 100000
lxc.cgroup.cpu.shares = 1024
lxc.rootfs = /zfspve/subvol-100-disk-1
lxc.network.type = veth
lxc.network.veth.pair = veth100i0
lxc.network.hwaddr = 32:37:62:63:38:36
lxc.network.name = eth0

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 11:02:41AM -0800, jcarnus wrote:
> root@debian:~# cat /var/lib/lxc/100/config
> lxc.arch = amd64
> lxc.include = /usr/share/lxc/config/debian.common.conf
> lxc.console = none
> lxc.cgroup.devices.deny = c 5:1 rwm
> lxc.tty = 0
> lxc.environment = TERM=linux
> lxc.utsname = pbx
> lxc.cgroup.memory.limit_in_bytes = 536870912
> lxc.cgroup.memory.memsw.limit_in_bytes = 1073741824
> lxc.cgroup.cpu.cfs_period_us = 100000
> lxc.cgroup.cpu.cfs_quota_us = 100000
> lxc.cgroup.cpu.shares = 1024
> lxc.rootfs = /zfspve/subvol-100-disk-1
> lxc.network.type = veth
> lxc.network.veth.pair = veth100i0
> lxc.network.hwaddr = 32:37:62:63:38:36
> lxc.network.name = eth0

Huh, that looks ok. Can you paste exactly the commands you're using to
dump and restore? Also, what version of liblxc are you on?

**Comment by jcarnus** [comment]
The commands :
root@debian:~/ckt# lxc-start -n 100
root@debian:~/ckt# lxc-checkpoint -n 100 -D ./ -v
root@debian:~/ckt# lxc-stop -n 100
root@debian:~/ckt# lxc-checkpoint -r -n 100 -D ./ -v

Btw, it not possible to do a lxc-checkpoint on a freeze container. Never return....

root@debian:/usr# ldconfig  -v | grep liblxc
        liblxc.so.1 -> liblxc.so.1.1.4

**Comment by tych0** [comment]
On Thu, Nov 05, 2015 at 11:15:35AM -0800, jcarnus wrote:
> The commands :
> root@debian:~/ckt# lxc-start -n 100
> root@debian:~/ckt# lxc-checkpoint -n 100 -D ./ -v
> root@debian:~/ckt# lxc-stop -n 100
> root@debian:~/ckt# lxc-checkpoint -r -n 100 -D ./ -v

Ok, I think I just realized what was going on. The restore code
just looks for a bridge in your config file, but doesn't use a default
bridge. Looks like in your config you don't have a bridge specified.
If you specify:

lxc.network.link = lxcbr0 # or whatever

you should be able to restore.

> Btw, it not possible to do a lxc-checkpoint on a freeze container. Never return....

Huh, ok. I'll take a look.

**Comment by jcarnus** [comment]
I don't use bridge.
Here /etc/network/interfaces.
auto lo
iface lo inet loopback
allow-hotplug eth0
iface eth0 inet dhcp
auto vmbr0
iface vmbr0 inet dhcp
        bridge_ports eth0
        bridge_stp off
        bridge_fd 0

If I add 
lxc.network.link = vmbr0, I got a 545 network is not created for 'lxc.network.link' when lxc-start.

And network is fully functionnal inside lxc ct

**Comment by tych0** [comment]
Right, the current lxc-checkpoint requires you to have one, unfortunately. I've sent a patch to fix this just now:

https://lists.linuxcontainers.org/pipermail/lxc-devel/2015-November/012727.html

and we are planning to tag 1.1.5 on Monday, so this should be fixed soon.

**Comment by jcarnus** [comment]
Ok, so for me, the restore issue if different than the dump issue of this original thread. Dump is now fine

**Comment by jcarnus** [comment]
I just test with lxc 1.1.5 and restore seems ok now.
Thanks a lot

**Comment by xemul** [comment]
I'll keep this open for a while till we fix "extra tasks in freeze cgroup" issue in CRIU. ( @avagin )

**Comment by xemul** [comment]
Fix is here 391efc9d7ef4ba7020f1705db7834a7a998d40aa
